contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -900,8 +895,130 @@ clearly described and the permissions issues should be resolved.
 This could be done with a double opt-in (an opt-in from both the referrer and the refer-ee) or with secondary permissions
 checks performed in admission. 
 
-TODO: Plugins, extensions, nested kinds, headers
+### Naming of the reference field
+
+The name of the reference field should be of the format "{field}Ref", with "Ref" always included in the suffix.
+
+The "{field}" component should be named to indicate the purpose of the reference. For example, "targetRef" in an
+endpoint indicates that the object reference specifies the target.
+
+It is okay to have the "{field}" component indicate the resource type. For example, "secretRef" when referencing
+a secret. However, this comes with the risk of the field being a misnomer in the case that the field is expanded to
+reference more than one type.
+
+### Referencing resources with multiple versions
+
+Most resources will have multiple versions. For example, core resources
+will undergo version changes as it transitions from alpha to GA.
+
+Controllers should assume that a version of a resource may change, and include appropriate error handling.
+
+### Handling of resources that do not exist
+
+There are multiple scenarios where a desired resource may not exist. Examples include:
+
+- the desired version of the resource does not exist.
+- race condition in the bootstrapping of a cluster resulting a resource not yet added.
+- user error.
+
+Controllers should be authored with the assumption that the referenced resource may not exist, and include
+error handling to make the issue clear to the user.
+
+### Object References Examples
+
+The following sections illustrate recommended schemas for various object references scenarios.
+
+The schemas outlined below are designed to enable purely additive fields as the types of referencable
+objects expand, and therefore are backwards compatible.
+
+For example, it is possible to go from a single resource type to multiple resource types without
+a breaking change in the schema.
+
+#### References to a single resource type 
+
+A single kind object reference is straightforward in that the controller can hard-code most qualifiers needed to identify the object. As such as the only value needed to be provided is the name (and namespace, although cross-namespace references are discouraged):
```

Comment:
hm, I thought we had agreed that group and kind should be included in this case too, for the purpose of future-compatibility (in case more kinds are added eventually) and copy-paste error prevention?

Reply:
For a single resource type, that feels a little overkill for me. I understood the agreement as including group for multiple resources (even what not necessary to uniquely qualify for the controller) due to copy-paste resulting in a different controller interpreting the "resource" as a resource for a different group.

For a single resource, it feels a bit verbose. It means adding two lines of boilerplate to every reference.

I would think copy-paste in a single kind reference wouldn't be common, but I don't have any anecdotal evidence for or against.

Reply:
> Maybe we can add a sentence something like, `If there's a chance this reference could eventually have more than one target kind of resource, it should be treated as a multi-resource-type reference from the beginning, because it's hard to be backwards compatible while adding group and kind fields.`?

Sounds good! Makes sense to educate and coach the implementer to design their references to be future compatible and consider the design ramifications. I'll get an update out when I have a chance.

Reply:
added!